source/agora/common/BlockStorage.d

```diff
@@ -0,0 +1,277 @@
+
+    /// Index is block height, Value is position of file
+    /// We will change to index file in the next step.
+    private size_t[] block_positions;
+
+    /// Ctor
+    public this (string path)
+    {
+        this.is_mapped = false;
+        this.path = path;
+    }
+
+    /***************************************************************************
+
+        Make a file name using the index.
+
+        Returns:
+            Returns the file name.
+
+    ***************************************************************************/
+
+    private string getFileName ()
+    {
+        return buildPath(this.path, "block.dat");
+    }
+
+    /***************************************************************************
+
+        Open memory mapped file.
+
+        If a size change occurs, open it again.
+
+        Params:
+            resize = The size of the file to change. If 0 does not change it.
+
+    ***************************************************************************/
+
+    private void map (size_t resize = 0)
+    {
+        if (this.is_mapped && (resize == 0))
+            return;
+
+        if (this.is_mapped)
+            this.unMap();
+
+        this.file =
+            new MmFile(
+                this.getFileName(),
+                MmFile.Mode.readWrite,
+                resize,
+                null
+            );
+
+        this.is_mapped = true;
+    }
+
+    /***************************************************************************
+
+        Close memory mapped file.
+
+    ***************************************************************************/
+
+    private void unMap ()
+    {
+        if (!this.is_mapped)
+            return;
+        import core.memory : GC;
+
+        destroy(this.file);
+        GC.free(&this.file);
+        this.is_mapped = false;
+    }
+
+    /***************************************************************************
+
+        Save block to the file.
+
+        Params:
+            block = `Block` to save
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool saveBlock (const ref Block block)
+    {
+        if (block.header.height != this.block_positions.length)
+            return false;
+
+        const ubyte[] serialized_block = serializeFull(block);
+
+        const size_t bidx = block.header.height;
+        size_t pos;
+
+        // first in this file
+        if (bidx == 0)
+        {
+            pos = 0;
+            // resize to serialized_block.length
+            this.map(serialized_block.length);
+        }
+        else
+        {
+            this.map();
+            pos = file.length;
+            // increase size by serialized_block.length
+            this.map(file.length + serialized_block.length);
+        }
+
+        // add information to the look up table.
+        if (this.block_positions.length < block.header.height + 1)
```

Comment:
Shouldn't this always be true? Because you have the check above `if (block.header.height != this.block_positions.length) return false`. So you can do:

```D
this.block_positions ~= pos;
```

Reply:
That's a good point. I'll apply it.